Client/Ecosia/UI/WhatsNew/DataProvider/Version.swift

```diff
@@ -0,0 +1,124 @@
+    init?(_ versionString: String) {
+        let components = versionString.split(separator: ".")
+        guard components.count == 3,
+              let major = Int(components[0]),
+              let minor = Int(components[1]),
+              let patch = Int(components[2]) else {
+            return nil
+        }
+        
+        self.major = major
+        self.minor = minor
+        self.patch = patch
+    }
+        
+    /// A string representation of the `Version`.
+    var description: String {
+        return "\(major).\(minor).\(patch)"
+    }
+}
+
+extension Version: Comparable {
+    
+    /// Compares two `Version` instances for equality.
+    ///
+    /// - Parameters:
+    ///   - lhs: A `Version`.
+    ///   - rhs: Another `Version`.
+    ///
+    /// - Returns: `true` if both instances represent the same version, `false` otherwise.
+    static func ==(lhs: Version, rhs: Version) -> Bool {
+        return lhs.major == rhs.major && lhs.minor == rhs.minor && lhs.patch == rhs.patch
+    }
+    
+    /// Compares two `Version` instances to determine their ordering.
+    ///
+    /// - Parameters:
+    ///   - lhs: A `Version`.
+    ///   - rhs: Another `Version`.
+    ///
+    /// - Returns: `true` if the instance on the left should come before the one on the right, `false` otherwise.
+    static func <(lhs: Version, rhs: Version) -> Bool {
+        if lhs.major != rhs.major {
+            return lhs.major < rhs.major
+        }
+        if lhs.minor != rhs.minor {
+            return lhs.minor < rhs.minor
+        }
+        return lhs.patch < rhs.patch
+    }
+}
+
+extension Version: Hashable {
+    
+    /// Adds this value to the given hasher.
+    ///
+    /// - Parameter hasher: The hasher to use when combining the components of this instance.
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(major)
+        hasher.combine(minor)
+        hasher.combine(patch)
+    }
+}
+
+/// Extension handling previous version retrieval and saving current version.
+extension Version {
+    
+    /// Retrieve the previously saved version or save the current version if none exists.
+    ///
+    /// This method tries to retrieve a previously saved version from `UserDefaults`.
+    /// If no version has been saved before, it saves the `current` version passed into the function.
+    ///
+    /// - Parameters:
+    ///   - current: The current `Version` to save if no version exists. Optional.
+    ///   - prefs: The `UserDefaults` instance to use for saving and retrieving the version.
+    ///            Defaults to `UserDefaults.standard`.
+    ///
+    /// - Returns: The previously saved `Version` if it exists, otherwise returns the current version.
+    ///            Returns `nil` if both the saved version and the current version are `nil`.
+    static func retrievePreviousVersionElseSaveCurrent(_ current: Version?, using prefs: UserDefaults = UserDefaults.standard) -> Version? {
```

Comment:
Might have misunderstood, but isn't this always retrieving the current version instead of the previous one like the name and description suggest?
I guess the only case it return the previous is if both match 🤔 

I like the idea but found it a bit confusing to be honest... In my opinion this storing logic should not belong to `Version` itself, but to the `DataProvider`. For example `WhatsNewLocalDataProvider` would have a stored UserDefaults value like `"lastAppVersionWhereLocalWhatsNewWasShown"` and would return the items between those two versions (similar logic like you already have) and then save the next one. This is mostly due to the fact that, enabling multiple providers to consume from `Version`, we would be risking that one retrieves before the other and updates the shared stored object, which would make the second one to think the app has not updated.

I also understand I might be overcomplicating things since there should always be only one `WhatsNewDataProvider` for our view model, this is just something that comes to mind with this current concept. Just let me know if you think this is not relevant for our context.

Sorry for the long text as well, hope I managed to convey the idea and that this does not result in a long discussion topic 🙈  I'm available to chat about it as well 🙂

Reply:
I’d love to catchup on that sync as I would make sure I’ve got it 💯 right! Will schedule a meeting same day as planning session to not interrupt any focus flow. Thanks for pointing out! 🙏